### PR TITLE
Remove intermediate base64-ing in attachments

### PR DESCRIPTION
**Pull request by xmo-odoo**

### CODE REVIEW

**Comment by xmo-odoo** [comment]
to redo: move modules to using the `raw` attribute in case they're currently base64-encoding just to create their attachment.

File: odoo/addons/base/models/ir_actions_report.py [hunk 1]
```diff
@@ -212,7 +212,8 @@ def postprocess_pdf_report(self, record, buffer):
             return None
         attachment_vals = {
             'name': attachment_name,
-            'datas': base64.encodebytes(buffer.getvalue()),
+            'raw': buffer.getvalue(),
+            'datas_fname': attachment_name,
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_actions_report.py]:
```suggestion
            'raw': buffer.getvalue(),
```

File: odoo/addons/base/models/ir_attachment.py [hunk 2]
```diff
@@ -65,8 +65,7 @@ def force_storage(self):
         }[self._storage()]
 
         for attach in self.search(domain):
-            # pass mimetype, to avoid recomputation
-            attach.write({'datas': attach.datas, 'mimetype': attach.mimetype})
+            attach.write({'raw': attach.raw})
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py]:
```suggestion
            attach.write({'raw': attach.raw, 'mimetype': attach.mimetype})
```

File: odoo/addons/base/models/ir_attachment.py [hunk 3]
```diff
@@ -97,22 +96,17 @@ def _get_path(self, bin_data, sha):
         return fname, full_path
 
     @api.model
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py]:
```suggestion
```

File: odoo/addons/base/models/ir_attachment.py [hunk 4]
```diff
@@ -192,18 +186,37 @@ def _file_gc(self):
...
-                attach.datas = attach.db_datas
+                attach.raw = attach.db_datas
+
+    def _inverse_raw(self):
+        self._set_data(lambda a: a.raw or b'')
 
     def _inverse_datas(self):
+        self._set_data(lambda attach: base64.b64decode(attach.datas) if attach.datas else b'')
+
+    def _set_data(self, asbytes):
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py]: Maybe rename to set_attachment_data or something?

File: odoo/addons/base/models/ir_actions_report.py (hunk 1 above)
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_actions_report.py]:
```suggestion
            'raw': buffer.getvalue(),
```

File: odoo/addons/base/models/ir_actions_report.py [hunk 5]
```diff
@@ -212,7 +212,8 @@ def postprocess_pdf_report(self, record, buffer):
...
             'res_model': self.model,
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_actions_report.py]:
```suggestion
            'raw': buffer.getvalue(),
            'res_model': self.model,
```

File: odoo/addons/base/models/ir_attachment.py (hunk 2 above)
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py]:
```suggestion
            attach.write({'raw': attach.raw, 'mimetype': attach.mimetype})
```

File: odoo/addons/base/models/ir_attachment.py (hunk 4 above)
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py]:
```suggestion
    def _inverse_raw(self):
        self._set_attachment_data(lambda a: a.raw or b'')

    def _inverse_datas(self):
        self._set_attachment_data(lambda attach: base64.b64decode(attach.datas) if attach.datas else b'')

    def _set_attachment_data(self, asbytes):
```

File: addons/l10n_latam_invoice_document/models/account_move.py [hunk 6]
```diff
@@ -179,5 +179,5 @@ def _check_unique_vendor_number(self):
     def unlink(self):
         """ When using documents, on vendor bills the document_number is set manually by the number given from the vendor,
         the odoo sequence is not used. In this case We allow to delete vendor bills with document_number/move_name """
-        self.filtered(lambda x: x.type in x.get_purchase_types() and x.state in ('draft', 'cancel') and x.l10n_latam_use_documents).write({'name': '/'})
+        self.filtered(lambda x: x.is_purchase_document() and x.state in ('draft', 'cancel') and x.l10n_latam_use_documents).write({'name': '/'})
```
Comment (xmo-odoo) [inline addons/l10n_latam_invoice_document/models/account_move.py]: TODO: remove this fixup commit (properly moved to a separate PR & merged there)

File: odoo/addons/base/models/ir_attachment.py [hunk 7]
```diff
@@ -192,18 +186,37 @@ def _file_gc(self):
...
+                attach.raw = self._file_read(attach.store_fname)
             else:
-                attach.datas = attach.db_datas
+                attach.raw = attach.db_datas
+
+    def _inverse_raw(self):
+        self._set_attachment_data(lambda a: a.raw or b'')
 
     def _inverse_datas(self):
+        self._set_attachment_data(lambda attach: base64.b64decode(attach.datas) if attach.datas else b'')
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py:212]:
```suggestion
        self._set_attachment_data(lambda attach: base64.b64decode(attach.datas or b''))
```

File: odoo/addons/base/models/ir_attachment.py [hunk 8]
```diff
@@ -521,7 +539,8 @@ def create(self, vals_list):
                 values.pop(field, False)
             values = self._check_contents(values)
             if 'datas' in values:
-                values.update(self._get_datas_related_values(values.pop('datas'), values['mimetype']))
+                data = values.pop('datas')
+                values.update(self._get_datas_related_values(base64.b64decode(data) if data else b'', values['mimetype']))
```
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_attachment.py:543]:
```suggestion
                values.update(self._get_datas_related_values(base64.b64decode(data or b''), values['mimetype']))
```

**Comment by xmo-odoo** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo r-

**Comment by rco-odoo** [comment]
(Interrupted merge because nobody else reviewed it.)

I just had a quick look, and the proposal looks fine to me.

But I wonder about migrations: what about existing base64-encoded files in the filestore?
Neither the task description nor this pull request seem to address this potential issue.
Have you tested on some existing database?

**Comment by xmo-odoo** [comment]
> But I wonder about migrations: what about existing base64-encoded files in the filestore?
> Neither the task description nor this pull request seem to address this potential issue.
> Have you tested on some existing database?

It should not be changing anything storage-wise, it just removes intermediate encode/decode, and provides an API so we can remove more in the future: there's a bunch of places where we get binary data and encode it just so we can set it as the attachment's datas where it's immediately decoded back to raw. Initially this was part of this PR but it generates conflicts basically every day so I removed it to get the "core" merged.

**Comment by rco-odoo** [comment]
I just re-read everything and now things are clearer.  The following invariant still holds:
 - the filestore contains the raw data;
 - the field `datas` corresponds to the base64-encoded data.

The difference is that we avoid base64-encoding when not necessary, and we can access the raw data (without intermediate base64 encoding) with the field `raw`.

@robodoo r+
